ksp_plugin_adapter/utf8_marshaler.cs

```diff
@@ -46,62 +36,51 @@ internal class InUTF8Marshaler : UTF8Marshaler {
     }
     int size = utf8_.GetByteCount(value);
     IntPtr buffer = Marshal.AllocHGlobal(size + 1);
-    while (bytes_.Length < size + 1) {
-      bytes_ = new byte[2 * bytes_.Length];
-    }
-    utf8_.GetBytes(value, 0, value.Length, bytes_, 0);
-    bytes_[size] = 0;
-    Marshal.Copy(bytes_, 0, buffer, size + 1);
+    byte[] bytes = new byte[size + 1];
+    utf8_.GetBytes(value, 0, value.Length, bytes, 0);
+    bytes[size] = 0;
+    Marshal.Copy(bytes, 0, buffer, size + 1);
+    UnityEngine.Debug.LogError("UnownedUTF8Marshaler.MarshalManagedToNative " +
+                               GetHashCode() + " " + buffer + " " +
+                               managed_object);
     return buffer;
   }
 
-  public override object MarshalNativeToManaged(IntPtr native_data) {
-    throw Log.Fatal("use |OutUTF8Marshaler| for out parameters");
-  }
-
-  private static readonly InUTF8Marshaler instance_ = new InUTF8Marshaler();
-  private byte[] bytes_ = new byte[1];
-}
-
-// A marshaler for out parameter or return value UTF-8 strings whose ownership
-// is not taken by the caller.
-internal class OutUTF8Marshaler : UTF8Marshaler {
-  public static ICustomMarshaler GetInstance(string s) {
-    return instance_;
-  }
-
-  public override void CleanUpNativeData(IntPtr native_data) {}
-  public override IntPtr MarshalManagedToNative(object managed_object) {
-    throw Log.Fatal("use |InUTF8Marshaler| for in parameters");
-  }
-
   public override object MarshalNativeToManaged(IntPtr native_data) {
     int size;
     for (size = 0; Marshal.ReadByte(native_data, size) != 0; ++size) {}
-    while (bytes_.Length < size) {
-      bytes_ = new byte[2 * bytes_.Length];
-    }
-    Marshal.Copy(native_data, bytes_, 0, size);
-    return utf8_.GetString(bytes_, 0, size);
+    byte[] bytes = new byte[size];
+    Marshal.Copy(native_data, bytes, 0, size);
+    var result = utf8_.GetString(bytes, 0, size);
+    UnityEngine.Debug.LogError("UnownedUTF8Marshaler.MarshalNativeToManaged " +
+                               GetHashCode() + " " + native_data + " " +
+                               result);
+    return result;
   }
 
-  private static readonly OutUTF8Marshaler instance_ = new OutUTF8Marshaler();
-  private byte[] bytes_ = new byte[1];
+  private static readonly UnownedUTF8Marshaler instance_ =
+      new UnownedUTF8Marshaler();
 }
 
-// A marshaler for out parameter or return value UTF-8 strings whose ownership
-// is taken by the caller.
-internal class OutOwnedUTF8Marshaler : OutUTF8Marshaler {
+// A marshaler for UTF-8 strings whose ownership is taken from C++.  Useful for
+// out parameters and returned values.
+internal class OwnedUTF8Marshaler : UnownedUTF8Marshaler {
```

Comment:
By symmetry, `CededUTF8Marshaler`.

Note that deriving from the other one seems a bit weird now.

If we wanted to use it for in parameters, it would have to allocate via an interface function. We have no use for that at the moment, so it should probably `Log.Fatal` if we accidentally use it that way.

Reply:
Granted, inheritance is a bit odd, but then composition doesn't work so well for classes that cannot have nonstatic members.

I don't understand the part about allocating via an interface.

Reply:
Done.